### PR TITLE
[fix]: [PL-40664]: Adds processing time metrics for user group events

### PR DESCRIPTION
**Pull request by ashishsanodia-harness**
## Describe your changes
Access control listens user group events changes for ACL processing. With this PR, we aim to introduce processing time metrics for user group events in the access control service. These metrics will be used to create Grafana dashboard, enabling us to closely monitor the time taken for user group processing. If necessary, alerts can be configured accordingly.

Tests:
- Sending event processing time metrics for usergroup events in access control
Screenshot 2023-10-16 at 5 06 08 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54144)

### CODE REVIEW

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
trigger ut1, ut5, ut8

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
trigger ut4, ut6, ut9

File: access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java [hunk 1]
```diff
@@ -65,6 +74,12 @@ public boolean handle(Message message) {
       } else {
         harnessUserGroupService.sync(stripToNull(entityChangeDTO.getIdentifier().getValue()), scope);
       }
+      long eventProcessingTime = System.currentTimeMillis() - eventReadTime;
+      try (AccessControlMetricsContext context =
+               new AccessControlMetricsContext(stripToNull(entityChangeDTO.getIdentifier().getValue()),
+                   USER_GROUP_ENTITY_TYPE, getEventType(message), ACCESS_CONTROL_SERVICE.getServiceId())) {
+        metricService.recordDuration(ACCESS_CONTROL_ENTITY_PROCESSING_TIME, ofMillis(eventProcessingTime));
```
Comment (ankit-harness) [inline access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java:81]: Should we include processing time for events whose processing failed as well?

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
*[`access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java` line 81 at r3](https://reviewable.io/reviews/harness/harness-core/54144#gh-1360587103:-NgsctWC5JmL54lqrXEE:b-5eaxt4) ([raw file](https://github.com/harness/harness-core/blob/42348d5556ab3297e0bdf9913ba512307fda4cf5/access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java#L81)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Should we include processing time for events whose processing failed as well?
</blockquote></details>

Moved user group event metrics sending to finally block

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54144)* status: 6 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java` line 54 at r4](https://reviewable.io/reviews/harness/harness-core/54144#-NgsdTu04dfH5F8vXOPT:-NgsdTu04dfH5F8vXOPU:b-kdmo29) ([raw file](https://github.com/harness/harness-core/blob/10de2b00b1d01bed7dbeb7f26238d20ca1bd951b/access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java#L54)):*
> ```Java
>   public boolean handle(Message message) {
>     EntityChangeDTO entityChangeDTO = null;
>     long eventReadTime = System.currentTimeMillis();
> ```

We should use System.nanoTime() to calculate the time diff
https://stackoverflow.com/questions/351565/system-currenttimemillis-vs-system-nanotime

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
*[`access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java` line 54 at r4](https://reviewable.io/reviews/harness/harness-core/54144#-NgsdTu04dfH5F8vXOPT:-NgseqLj67xjXw8TbgWh:b-886gsx) ([raw file](https://github.com/harness/harness-core/blob/10de2b00b1d01bed7dbeb7f26238d20ca1bd951b/access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java#L54)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

We should use System.nanoTime() to calculate the time diff
https://stackoverflow.com/questions/351565/system-currenttimemillis-vs-system-nanotime
</blockquote></details>

Fixed

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54144)* status: 6 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java` line 85 at r5](https://reviewable.io/reviews/harness/harness-core/54144#-Ngsfbth2RZSjHJs4F_6:-Ngsfbth2RZSjHJs4F_7:b-o8ifvk) ([raw file](https://github.com/harness/harness-core/blob/c2b0d42e043051b4ae4563e127967c216a48d831/access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java#L85)):*
> ```Java
>                new AccessControlMetricsContext(stripToNull(entityChangeDTO.getIdentifier().getValue()),
>                    USER_GROUP_ENTITY_TYPE, getEventType(message), ACCESS_CONTROL_SERVICE.getServiceId())) {
>         metricService.recordDuration(ACCESS_CONTROL_ENTITY_PROCESSING_TIME, ofMillis(eventProcessingTime));
> ```

TimeUnit.NANOSECONDS.toMillis(eventProcessingTime);

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
*[`access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java` line 85 at r5](https://reviewable.io/reviews/harness/harness-core/54144#-Ngsfbth2RZSjHJs4F_6:-Ngsh__r6K87_66Wzo3N:b-886gsx) ([raw file](https://github.com/harness/harness-core/blob/c2b0d42e043051b4ae4563e127967c216a48d831/access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/events/UserGroupEventHandler.java#L85)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

TimeUnit.NANOSECONDS.toMillis(eventProcessingTime);
</blockquote></details>

Fixed

**Review by ankit-harness — COMMENTED** [review]
> Reviewed 4 of 6 files at r3, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54144)* status: 6 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by ankit-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54144)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Pull request merged**